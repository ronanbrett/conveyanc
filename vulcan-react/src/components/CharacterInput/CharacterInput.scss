@import "src/styles/vars";

.CharacterInput {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;

  .TextInput--single {
    margin-right: var(--spacing-xs);
    margin-bottom: var(--spacing-s);

    display: flex;
    justify-content: center;
    align-items: center;

    width: 32px;
    height: 64px;
    padding: 0;
    flex: 0 0 50px;

    input {
      padding: 0;
      text-align: center;
    }

    @include mq($until: wide) {
      width: 24px;
      flex: 0 0 50px;
    }

    &:nth-of-type(3) {
      margin-right: var(--spacing-l);
    }
  }
}
