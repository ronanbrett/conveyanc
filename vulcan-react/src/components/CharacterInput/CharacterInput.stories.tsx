import React, { useState } from "react";

import CharacterInput from "./CharacterInput";

import { action } from "@storybook/addon-actions";

export default {
  title: "Components/CharacterInput",
  decorators: [],
  component: CharacterInput,
};

export const Default = (props) => {
  return <CharacterInput size={7} value="D08XKT1" {...props}></CharacterInput>;
};

export const WithErrors = (props) => {
  const [errors, setErrors] = useState<string[]>();

  return (
    <div>
      <CharacterInput
        errors={errors}
        size={7}
        value="D08XKT1"
        {...props}
      ></CharacterInput>

      <button onClick={() => setErrors(["newError"])}>Add Error</button>
    </div>
  );
};
