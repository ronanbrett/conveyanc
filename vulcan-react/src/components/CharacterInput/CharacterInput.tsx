import { TextInput } from "@components";
import { FormContext } from "@core/contexts/FormContext";
import Keyboard from "components/Keyboard";
import { motion, useAnimation } from "framer-motion";
import React, {
  FC,
  SyntheticEvent,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { SyntheticEventData } from "react-dom/test-utils";
import "./CharacterInput.scss";

export interface CharacterInputProps {
  name?: string;
  value?: string;
  onChange?: (value?: string) => void;
  onUpdate?: (value?: string) => void;
  errors?: any[];
  size: number;
  children?: any;
}

const CharacterInput: FC<CharacterInputProps> = ({
  value: valueProp,
  onUpdate,
  name,
  errors,
  size,
  ...props
}) => {
  const [parts, setParts] = useState([]);
  const controls = useAnimation();
  const formContext = useContext(FormContext);

  const [value, setValue] = formContext.useFormInput(name, valueProp);

  const elRefs = React.useRef<HTMLInputElement[]>([]);

  useMemo(() => {
    if (!valueProp) {
      return setParts(new Array(size).fill(""));
    }

    if (valueProp.length > size) {
      throw new Error("Value doesnt fit");
    }

    setParts(valueProp.split(""));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useMemo(() => {
    if (valueProp === "") {
      setParts(new Array(size).fill(""));
    }
  }, [valueProp, size]);

  useEffect(
    () => {
      const value = parts.join("");
      setValue(value);
      onUpdate(value);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [parts]
  );

  useEffect(
    () => {
      if (errors?.length > 0) {
        triggerErrors();
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [errors]
  );

  const validate = (char: string, index: number) => {
    const match = RegExp(/A-Za-z0-9/);
    return match.test(char);
  };

  const onChange = (e: any, index: number) => {
    const value = e.target.value as string;
    validate(value, index);

    const newParts = [...parts];
    newParts.splice(index, 1, value.toUpperCase());
    setParts(newParts);

    if (index === size - 1 && value) {
      triggerValidation();
    }
  };

  const onKeyPress = (e: any, index: number) => {
    if (elRefs && elRefs.current[index + 1]) {
      setTimeout(() => {
        elRefs.current[index + 1].focus();
        elRefs.current[index + 1].select();
      });
    }
  };

  const onClick = (index: number) => {
    elRefs.current[index].select();
  };

  const onBackspace = (value: string, index: number) => {
    if (value) {
      return;
    }
    if (elRefs && elRefs.current[index - 1]) {
      elRefs.current[index - 1].focus();
      elRefs.current[index - 1].select();
    }
  };

  const onPaste = (evt: any) => {
    evt.clipboardData?.items[0]?.getAsString((text: string) => {
      let arr = [...parts];
      const txt = text.replaceAll(" ", "").toUpperCase().split("");
      arr = arr.map((val, index) => {
        return txt[index] || "";
      });
      setParts(arr);

      if (txt.length === size) {
        setTimeout(() => {
          triggerValidation();
        });
      }
    });
  };

  const onEnter = () => {
    triggerValidation();
  };

  const triggerValidation = () => {
    controls.set("inactive");
    controls.start("active");
  };

  const triggerErrors = () => {
    controls.set("inactive");
    controls.start("errors");
  };

  const onEsc = () => {
    setParts(new Array(size).fill(""));
    elRefs.current[0].focus();
  };

  const list = {
    inactive: {},
    active: {
      transition: {
        ease: "easeOut",
        staggerChildren: 0.07,
        staggerDirection: -1,
      },
    },
    errors: {
      transition: {
        ease: "easeOut",
        staggerChildren: 0.07,
        staggerDirection: 1,
      },
    },
  };

  return (
    <div>
      <motion.div
        initial="inactive"
        animate={controls}
        className="CharacterInput"
        variants={list}
      >
        {parts.length > 0 ? (
          parts.map((val, index, ref) => (
            <Keyboard
              key={index}
              onEnter={onEnter}
              onEsc={onEsc}
              onBackspace={(evt: SyntheticEventData) =>
                onBackspace(elRefs.current[index].value, index)
              }
            >
              <TextInput
                key={index}
                onPaste={onPaste}
                ref={(el: HTMLInputElement) => (elRefs.current[index] = el)}
                pattern="[A-Za-z0-9]{1}"
                onClick={() => onClick(index)}
                onChange={(evt: SyntheticEvent) => onChange(evt, index)}
                onKeyPress={(evt: SyntheticEvent) => onKeyPress(evt, index)}
                maxLength={1}
                size="single"
                type="text"
                value={val}
              />
            </Keyboard>
          ))
        ) : (
          <div></div>
        )}
      </motion.div>
    </div>
  );
};

export default CharacterInput;
