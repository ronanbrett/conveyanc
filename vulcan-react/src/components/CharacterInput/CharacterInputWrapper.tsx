import React, { FC, SyntheticEvent, useEffect, useState } from "react";

import CharacterInput, { CharacterInputProps } from "./CharacterInput";
import { useField, FieldConfig } from "formik";

const CharacterInputWrapper = ({
  size,
  onChange: onChangeExt,
  ...props
}: Partial<CharacterInputProps> & FieldConfig) => {
  const [field, meta, helpers] = useField(props);

  const onChange = (val: string) => {
    helpers.setValue(val);
  };

  return (
    <CharacterInput
      onUpdate={onChange}
      size={size}
      {...field}
      {...props}
    ></CharacterInput>
  );
};

export default CharacterInputWrapper;
