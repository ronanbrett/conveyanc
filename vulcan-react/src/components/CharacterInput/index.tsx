import CharacterInput from "./CharacterInput";
import CharacterInputWrapper from "./CharacterInputWrapper";

export { CharacterInput, CharacterInputWrapper };
