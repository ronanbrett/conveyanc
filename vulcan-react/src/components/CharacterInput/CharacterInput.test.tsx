import React from "react";
import { create } from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import CharacterInput from "./CharacterInput";
import { Formik, Form } from "formik";
import CharacterInputWrapper from "./CharacterInputWrapper";

describe("CharacterInput", () => {
  it("renders without error", () => {
    const component = create(
      <Formik initialValues={{}} onSubmit={(x) => console.log(x)}>
        <Form>
          <label htmlFor="propertyType">Property Type</label>
          <CharacterInput size={4}></CharacterInput>
        </Form>
      </Formik>
    );

    expect(component).toMatchSnapshot();
  });

  describe("for the CharacterInput dom", () => {
    it("should render a series of input", () => {
      let { container } = render(
        <Formik initialValues={{ test: "" }} onSubmit={(x) => console.log(x)}>
          <Form>
            <label htmlFor="propertyType">Property Type</label>
            <CharacterInputWrapper name="test" size={4}></CharacterInputWrapper>
          </Form>
        </Formik>
      );
      const inputs = container.querySelectorAll("input");
      expect(inputs.length).toBe(4);
    });

    it("should update the form", () => {
      const mockCallback = jest.fn();
      let { container } = render(
        <Formik initialValues={{ test: "" }} onSubmit={mockCallback}>
          <Form>
            <label htmlFor="propertyType">Property Type</label>
            <CharacterInputWrapper name="test" size={4}></CharacterInputWrapper>
          </Form>
        </Formik>
      );

      expect(mockCallback).not.toBeCalled();
      const inputs = container.querySelectorAll("input");
      expect(inputs.length).toBe(4);
      inputs[0].value = "A";
      fireEvent.change(inputs[0]);
    });
  });
});
