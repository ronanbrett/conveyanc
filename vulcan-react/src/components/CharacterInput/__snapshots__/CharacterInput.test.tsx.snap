// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterInput renders without error 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <label
    htmlFor="propertyType"
  >
    Property Type
  </label>
  <div>
    <div
      className="CharacterInput"
      style={Object {}}
    >
      <div
        className="TextInput__input__container TextInput--single"
        style={
          Object {
            "borderColor": "var(--border-color)",
            "transform": "none",
          }
        }
      >
        <input
          autoComplete="off"
          className="TextInput__input TextInput--text"
          maxLength={1}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onPaste={[Function]}
          pattern="[A-Za-z0-9]{1}"
          type="text"
          value=""
        />
      </div>
      <div
        className="TextInput__input__container TextInput--single"
        style={
          Object {
            "borderColor": "var(--border-color)",
            "transform": "none",
          }
        }
      >
        <input
          autoComplete="off"
          className="TextInput__input TextInput--text"
          maxLength={1}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onPaste={[Function]}
          pattern="[A-Za-z0-9]{1}"
          type="text"
          value=""
        />
      </div>
      <div
        className="TextInput__input__container TextInput--single"
        style={
          Object {
            "borderColor": "var(--border-color)",
            "transform": "none",
          }
        }
      >
        <input
          autoComplete="off"
          className="TextInput__input TextInput--text"
          maxLength={1}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onPaste={[Function]}
          pattern="[A-Za-z0-9]{1}"
          type="text"
          value=""
        />
      </div>
      <div
        className="TextInput__input__container TextInput--single"
        style={
          Object {
            "borderColor": "var(--border-color)",
            "transform": "none",
          }
        }
      >
        <input
          autoComplete="off"
          className="TextInput__input TextInput--text"
          maxLength={1}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onPaste={[Function]}
          pattern="[A-Za-z0-9]{1}"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</form>
`;
